### PR TITLE
[DRUP-610] Add url as source for apidocs spec file and (DRUP-459) allow re-fetching it afterwards.

### PR DESCRIPTION
**Pull request by Jaesin**
Continuation of: https://github.com/apigee/apigee-edge-drupal/pull/191

### CODE REVIEW

File: composer.json
Comment (mxr576) [inline composer.json:8]: Require as `^1.0` and add [conflict](https://getcomposer.org/doc/04-schema.md#conflict) as `<=1.0-rc4` that way you do not need to change the minimum requirement when 1.0 is out.

File: src/Entity/ApiDoc.php
Comment (mxr576) [inline src/Entity/ApiDoc.php]: https://github.com/apigee/apigee-edge-drupal/pull/191/files#r283731935

image

File: src/Entity/ApiDoc.php
Comment (mxr576) [inline src/Entity/ApiDoc.php]: As I can see you are not calling `$this->t()` here this import is unnecessary.
(Because [it silently falls back](https://github.com/drupal/core/blob/8.7.x/lib/Drupal/Core/StringTranslation/StringTranslationTrait.php#L103-L105) to the global static service, it is also considered as bad practice using it in a context where DI is not available.)

File: src/Entity/Form/ApiDocReimportSpecForm.php
Comment (mxr576) [inline src/Entity/Form/ApiDocReimportSpecForm.php:113]: This is exactly why I suggested earlier to throw an exception when file fetching or saving fails instead of returning a boolean indicator value which is not the OOP solution.

File: src/SpecFetcher.php
Comment (mxr576) [inline src/SpecFetcher.php:137]: https://github.com/apigee/apigee-edge-drupal/pull/191/files#r284210698

image

File: src/SpecFetcher.php
Comment (mxr576) [inline src/SpecFetcher.php]: I do not think this solves the problem with internationalization. Please verify this.
You can easily do that by enabling internationalization and trying to translate `Error while saving API Doc spec file from URL on API Doc ID: %id. Error: %error` string.

File: src/Plugin/Validation/Constraint/ApiDocFileLinkConstraintValidator.php
Comment (mxr576) [inline src/Plugin/Validation/Constraint/ApiDocFileLinkConstraintValidator.php]: https://github.com/guzzle/guzzle/blob/master/CHANGELOG.md#600-beta1
> The exceptions request option has been deprecated in favor of the http_errors request options.

Also, I think this option is enabled by default: `\GuzzleHttp\Client::configureDefaults()`.

File: src/SpecFetcher.php
Comment (mxr576) [inline src/SpecFetcher.php:225]: https://github.com/apigee/apigee-edge-drupal/pull/191/files#r284212260
image

File: src/SpecFetcher.php
Comment (mxr576) [inline src/SpecFetcher.php:96]: https://github.com/apigee/apigee-edge-drupal/pull/191/files#r283729115
image

**Review by mxr576 — CHANGES_REQUESTED** [review]

**Comment by mxr576** [comment]
https://github.com/apigee/apigee-api-catalog-drupal/pull/2#issuecomment-493620759
The specfetcher service and the automated imported process still do not have a test coverage although these are the most important changes of this PR.

File: composer.json
Comment (Jaesin) [reply composer.json:8]: This is simpler the way it is. Won't we want to require `apigee_edge` > 1.0.0 when it is released?

File: src/Entity/ApiDoc.php
Comment (Jaesin) [reply src/Entity/ApiDoc.php]: I prefer to keep code/classes to a minimum whenever possible and creating a new class and overriding a storage handler just to use dependency injection design pattern seems unnecessary to me.

File: src/Entity/ApiDoc.php
Comment (Jaesin) [reply src/Entity/ApiDoc.php]: Good catch. I'll remove the trait.

File: src/Entity/Form/ApiDocReimportSpecForm.php
Comment (Jaesin) [reply src/Entity/Form/ApiDocReimportSpecForm.php]: Yes, this can be addressed in a followup.

File: src/SpecFetcher.php
Comment (Jaesin) [reply src/SpecFetcher.php:137]: Interesting use case, that could be changed if it is needed in testing.

File: src/SpecFetcher.php
Comment (Jaesin) [reply src/SpecFetcher.php]: OK. Good point. I will test it.

File: src/Plugin/Validation/Constraint/ApiDocFileLinkConstraintValidator.php
Comment (Jaesin) [reply src/Plugin/Validation/Constraint/ApiDocFileLinkConstraintValidator.php]: Thanks for spotting, I will look into this.

File: src/Entity/Form/ApiDocReimportSpecForm.php
Comment (Jaesin) [reply src/Entity/Form/ApiDocReimportSpecForm.php:113]: This TODO was added so it can be changed in a follow up task.

File: src/SpecFetcher.php
Comment (Jaesin) [reply src/SpecFetcher.php:225]: I have no problem with addressing this in a followup PR.

File: src/SpecFetcher.php
Comment (Jaesin) [reply src/SpecFetcher.php:96]: Although this is a micro-optimization, I have no problem addressing it in a followup task.

File: src/Entity/ApiDoc.php
Comment (mxr576) [reply src/Entity/ApiDoc.php]: Handling things in the right place it is never an unnecessary work :) It is not just about DI, when Drupal core maintainers finally deprecate and remove CRUD methods from entity classes then leaving this here will cause extra work for us that we could have eliminated before this PR got merged.
If we'll need to create a storage class for this entity type later, what ensures this code will be moved to its rightful place at that time? Nothing.

File: src/Entity/Form/ApiDocReimportSpecForm.php
Comment (mxr576) [reply src/Entity/Form/ApiDocReimportSpecForm.php:113]: Which is the related follow-up task to this? Probably it would be better to create it as a GH issue, because that could be referenced in this PR.
Also, please keep in mind that if this module gets stable before this issue gets fixed then technically, you should not change what it is in `SpecFetcherInterface::fetchSpec()` without releasing a new major version because that would be a BC breaking change. So any implementation that implements this interface should not throw any exception and should always return a boolean value because this is what it is in the contract.

File: src/SpecFetcher.php
Comment (mxr576) [reply src/SpecFetcher.php:96]: Which is the related follow-up task to this? Probably it would be better to create it as a GH issue, because that could be referenced in this PR.

File: src/SpecFetcher.php
Comment (mxr576) [reply src/SpecFetcher.php:225]: Which is the related follow-up task to this? Probably it would be better to create it as a GH issue, because that could be referenced in this PR.

File: composer.json
Comment (cnovak) [reply composer.json:8]: I agree to leave it as is, will let user install anything above rc4, why force to 1.0.0?

File: src/Entity/ApiDoc.php
Comment (cnovak) [inline src/Entity/ApiDoc.php]: Need to add t() to label and description for this field also

File: composer.json
Comment (mxr576) [reply composer.json:8]: `^1.0` only enforces >=1.0.0 when 1.0 is out. Otherwise, it allows installation of any non-stable releases. The conflict solves that is always minimum `1.0-rc4`. This is the constraint that is being suggested on every Drupal.org module/theme install pages: https://www.drupal.org/project/apigee_edge/releases/8.x-1.0-rc4
> Install with Composer: $ composer require 'drupal/apigee_edge:^1.0'

**If 1.0.0 is out** but you left `~1.0-rc4` in this composer.json, depending on the [minimum-stability](https://getcomposer.org/doc/04-schema.md#minimum-stability) configuration of the Drupal project, it could happen that still `1.0-rc4` gets installed on the project. Although, if you use `^1.0' as a constraint then always an >=1.0.0 release gets installed.

Otherwise, if you leave `1.0-rc4` now and 1.0.0 is out and you would like to avoid the above-described scenario you will have to bump the minimum required version again to `^1.0` and release a new version from the `apigee-api-catalog-drupal` module which could have been withdrawn by using what I am suggesting here.

I am fine if you merge this with the original constraint and later bump the minimum requirement later but sometimes it does not hurt if we think it through what are the consequences of a change and how the extra work caused by the change could be avoided

Comment (mxr576) [reply composer.json:8]: > Although, if you use `^1.0' as a constraint then always a >=1.0.0 release gets installed.

Well, I checked this in my lunchtime on http://semver.mwl.be and if it is not lying then it seems I expected more from Composer as I should have. It seems the `minimum-stability !== stable` can still allow installation of RC releases after the stable is out ;S

For me, it still makes sense to `require` the upcoming `^1.0` stable tag and the unstable minimum requirement (`<=rc4`) as a `conflict` because at least semantically it suggests for any reviewer that we prefer the stable release from the `apigee_edge` module and gives fewer credits to the unstable rc4 release that we wanna forget after stable is out. But this is my opinion only.

File: src/Entity/ApiDoc.php
Comment (Jaesin) [reply src/Entity/ApiDoc.php]: Drupal core uses this static service resolver ~825 times in runtime code. ~200 times in core libraries (as opposed to modules) so It's not a terrible thing.

File: src/SpecFetcher.php
Comment (Jaesin) [reply src/SpecFetcher.php]: Fixed.

File: src/Plugin/Validation/Constraint/ApiDocFileLinkConstraintValidator.php
Comment (Jaesin) [reply src/Plugin/Validation/Constraint/ApiDocFileLinkConstraintValidator.php]: Instead of setting `'http_errors' => TRUE` I'm just going to remove the option since that is the default value.

**Comment by Jaesin** [comment]
There are no new tests in this PR. Tests will be added in a followup PR.

Existing test results: 

```
Testing started at 17:47 ...
/usr/local/Cellar/php@7.2/7.2.15/bin/php .../vendor/phpunit/phpunit/phpunit --configuration .../phpunit.xml .../modules/contrib/apigee_api_catalog/tests --teamcity
PHPUnit 6.5.13 by Sebastian Bergmann and contributors.

Testing .../modules/contrib/apigee_api_catalog/tests


Time: 57.47 seconds, Memory: 6.00MB

OK (9 tests, 116 assertions)
```

File: src/Entity/Form/ApiDocReimportSpecForm.php
Comment (Jaesin) [reply src/Entity/Form/ApiDocReimportSpecForm.php:113]: https://github.com/apigee/apigee-api-catalog-drupal/issues/14

**Review by cnovak — APPROVED** [review]
I think this looks good now w/latest updates, should have issues in for anything outstanding.  After merging other commits this PR had conflicts.  I resolved merge conflicts, then ran phpcs and phpunit and all tests now pass.  

However, @Jaesin you should look over [ApiDocAccessControlHandler:checkAccess()](https://github.com/apigee/apigee-api-catalog-drupal/pull/2/files#diff-df4bc83028e8cf96656637de8abf3ed9L38) to make sure that method was properly merged.

Test results:

```
 ../vendor/bin/phpunit --configuration core modules/contrib/apigee_api_catalog/ 
PHPUnit 6.5.14 by Sebastian Bergmann and contributors.

Testing modules/contrib/apigee_api_catalog/
..............                                                    14 / 14 (100%)

Time: 3.36 minutes, Memory: 10.00MB

OK (14 tests, 158 assertions)
```

File: src/Entity/ApiDoc.php
Comment (mxr576) [reply src/Entity/ApiDoc.php]: Well, just because you see a pattern repeating in Drupal 8 core it does necessarily mean that is a good architectural solution. This is the reason why I learning from other systems and reading books about software-architecture and OOP coding to extend my current knowledge and do not just blindly trust what I see in Drupal core.
Let's take `final` classes as an example, you do not see in Drupal core and probably that is one of the reasons why BC breaking is so easy and sometimes upgrade path suffers from changes made in Drupal core on non-final classes or properties and methods that were public/protected.
Also do not forget the OOP was a huge switch compared with Drupal 7 for everyone, even for some Drupal maintainer. So the Drupal 8 core currently contains initial decisions that they made in the beginning and now they are trying to get rid of that by using deprecations. The problem is with deprecations that it does not allow to change everything that they want.

**Comment by Jaesin** [comment]
`ApiDocAccessControlHandler:checkAccess()` looks right to me and all of the access tests pass so it should be good to go.

**Pull request merged**
